### PR TITLE
Get all the VIPs of a service entry through MCP over xDS

### PR DESCRIPTION
**Pull request by zhaohuabing**
Related issue is here: https://github.com/istio/istio/issues/36694

Signed-off-by: zhaohuabing <huabingzhao@tencent.com>

**Please provide a description of this PR:**

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [X] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [X] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Comment by zhaohuabing** [comment]
I guess we could, but it would be confusing, because its role is not a proxy.

[Huabing Zhao - Chat @ Spike](https://spikenow.com/r/a/?ref=spike-organic-signature&_ts=1dmtlo)	[1dmtlo]

On January 13, 2022 at 13:50 GMT, Rama Chavali ***@***.***> wrote:

But this use case is another node type, which is a
"controller" running on the control plane.

Why cant we distinguish with proxy metadata?

File: pilot/pkg/model/context.go [hunk 1]
```diff
@@ -744,9 +744,11 @@ const (
 
 	// Router type is used for standalone proxies acting as L7/L4 routers
 	Router NodeType = "router"
+
+	ControllerNode NodeType = "controller"
```
[the remark this answers is not in the document]
Comment (zhaohuabing) [reply pilot/pkg/model/context.go:748]: Are there any concerns about introducing a new node type?

**Comment by zhaohuabing** [comment]
@costinm @hzxuzhonghu Could you please share your thoughts about this? Should we just use Proxy Metadata or introduce a "controller" node type?

**Comment by zhaohuabing** [comment]
If we choose to use MetaData, we have a few options: Labels, Annotations, Raw. Neither Labels nor Annotations fit the semantic of the use case. May I just define a key "isController" in the Raw map?

**Comment by ramaraochavali** [comment]
> May I just define a key "isController" in the Raw map?

My preference would be this. as introducing a new node type and not all code paths honouring that type (we mostly check if it is router or sidecar in majority of the code) seems incorrect to me. But I will wait for others feedback

**Comment by zhaohuabing** [comment]
> > May I just define a key "isController" in the Raw map?
> 
> My preference would be this. as introducing a new node type and not all code paths honouring that type (we mostly check if it is router or sidecar in majority of the code) seems incorrect to me. But I will wait for others feedback

@costinm @hzxuzhonghu  How do you think about it?

**Comment by hzxuzhonghu** [comment]
Option 2: there is a generator field in `NodeMetadata`, we can make use of it.

**Comment by zhaohuabing** [comment]
> Option 2: there is a generator field in `NodeMetadata`, we can make use of it.

I need to pull the MCP resources, so the generator field has already been set as "API".

**Comment by hzxuzhonghu** [comment]
`API` seems enough

**Comment by zhaohuabing** [comment]
> `API` seems enough

The Aeraki use case is that Aeraki needs to pull MCP resources(Service Entry, VIrtual Service, etc.) from the Istio, just the same as the "api" generator currently does, but from a "controller" point of view, so it needs all the VIPs in a Service Entry, not just the address of the same cluster of the "proxy".

**Comment by ramaraochavali** [comment]
Thank you. LGTM. Will leave it for others to review

**Comment by costinm** [comment]
I think we should look more at the feature we want - not at the particular role.

You seem to want a custom filtering of the resources that are returned. It's not specific to controllers.

We can start with a Result_filter=none metadata. Later we can refine it, or even add it the the resource name.

**Comment by zhaohuabing** [comment]
This approach is more generic and can be applied to more use cases, I'll change the code.

**Comment by hzxuzhonghu** [comment]
MCP client does not want to filter out some resources, instead, it wants to know all resources.  There is a lack of use cases to implement more logic in pilot to support generic filtering.

**Comment by hzxuzhonghu** [comment]
Put a hold until come to an agreement

**Comment by zhaohuabing** [comment]
Should we put all the addresses in the result by default? Will the behavior cause incompatible change?

**Comment by zhaohuabing** [comment]
@hzxuzhonghu @ramaraochavali @costinm Hi guys, is anything I can do to move this forward?

**Comment by stevenctl** [comment]
> Should we put all the addresses in the result by default? Will the behavior cause incompatible change?

I like the idea, but depends on how consumers are using this.. If they're just building things that match these VIPs, they'll have a few irrelevant ones, but there could easily be cases where traffic starts getting sent to VIPs that are invalid for a given cluster. 

Do consumers properly set `CLUSTER_ID` meta? They should... and maybe if that's unspecified `""` (_not_ the default `"Kubernetes"' value) we don't filter VIPs based on CLUSTER_ID?

**Comment by zhaohuabing** [comment]
> Do consumers properly set `CLUSTER_ID` meta? They should... and maybe if that's unspecified `""` (_not_ the default `"Kubernetes"' value) we don't filter VIPs based on CLUSTER_ID?

I think this approach is the best so far. It works for my use case and also has the least impacts on existing users.

@hzxuzhonghu @ramaraochavali @costinm if you also agree with this, I can change this PR this way.

**Comment by stevenctl** [comment]
Another idea is `cluster_id=*` – less likely to break other clients. Only issue is making that wildcard have consistent behavior in all the places that we use cluster_id..

File: pilot/pkg/model/service.go [hunk 2]
```diff
@@ -769,6 +780,20 @@ func (s *Service) GetAddressForProxy(node *Proxy) string {
...
 
+// getAllAddresses returns a Service's all addresses.
+func (s *Service) getAllAddresses() []string {
+	var addresses []string
+	addressMap := s.ClusterVIPs.GetAddresses()
+	for _, clusterAddresses := range addressMap {
+		addresses = append(addresses, clusterAddresses...)
+	}
+
+	if len(addresses) == 0 && s.DefaultAddress != "" {
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/service.go:791]: DefaultAddress  can be "0.0.0.0"? 
In this case, there maybe autoallocated address, do we need to return it?

Comment (zhaohuabing) [reply pilot/pkg/model/service.go:791]: > DefaultAddress can be "0.0.0.0"? In this case, there maybe autoallocated address, do we need to return it?

We shouldn't return “0.0.0.0" or the auto-allocated address.

**Review by hzxuzhonghu — APPROVED** [review]

**Pull request merged**